KAFKA-6761: Construct Physical Plan using Graph, Reduce streams footprint part III

bbejeck opened the pull request:
Sorry for the massive PR, but at this point, it's very difficult to break up into smaller parts now that we are building the logical and physical plan.
It's worth noting at the moment this PR does not include optimizations to help with the review burden, the 4th PR will include an optimization for repartition topics and re-using source topics as changelogs for KTables
 

The specific changes in this PR from the second PR include
2. Changed the types of graph nodes to names conveying more context
2. Build the entire physical plan from the graph, after `StreamsBuilder.build()` is called.

Other changes are addressed directly as review comments on the PR.

Testing consists of using all existing streams tests to validate building the physical plan with graph

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
Intercepting calls to InternalStreamsBuilder, and InternalTopologyBuilder to add to graph

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/BaseRepartitionNode.java:
```diff
@@ -0,0 +1,97 @@
+import org.apache.kafka.streams.processor.FailOnInvalidTimestamp;
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+
+public abstract class BaseRepartitionNode<K, V> extends StreamsGraphNode {
```
This class is created to contain common information for repartition operations.  There are two types of repartitioning currently.  When changing a key in a `KStream`s method (`map`, `flatMap` `selectKey` etc) or when performing a `KTable.groupBy` operation.  The former is eligible for optimization, while the `KTable.groupBy` is not.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/BaseRepartitionNode.java:
```diff
@@ -0,0 +1,97 @@
+        topologyBuilder.addSink(sinkName,
+                                repartitionTopic,
+                                keySerializer,
+                                getValueSerializer(),
```
There are 2 difference between the repartition operations the first is the `Serializer` and `Deserializer` required

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/BaseRepartitionNode.java:
```diff
@@ -0,0 +1,97 @@
+                                  repartitionTopic);
+
+        topologyBuilder.addProcessor(processorParameters.processorName(), processorParameters.processorSupplier(), getProcessorPredecessorName());
+
```
The second difference between repartition operations is the name used to wire up the repartitioning processor.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/GroupedTableOperationRepartitionNode.java:
```diff
@@ -0,0 +1,136 @@
+import org.apache.kafka.streams.kstream.internals.MaterializedInternal;
+
+public class GroupedTableOperationRepartitionNode<K, V> extends BaseRepartitionNode {
+
```
This class was created to represent the _**non optimizable**_ repartition resulting from a `KTable.groupBy` operation

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:
```diff
@@ -0,0 +1,171 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals.graph;
```
Same class just moved to `graph` package

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/OptimizableRepartitionNode.java:
```diff
@@ -0,0 +1,125 @@
+import org.apache.kafka.common.serialization.Serde;
+
+public class OptimizableRepartitionNode<K, V> extends BaseRepartitionNode {
+
```
This class represents a repartition operation that _**is eligible**_ to get optimized away.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorParameters.java:
```diff
@@ -15,7 +15,7 @@
 
-package org.apache.kafka.streams.kstream.internals;
+package org.apache.kafka.streams.kstream.internals.graph;
 
```
The class is the same just moved to a new package. This is the same for all graph objects, so I won't continue to repeat this comment.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java:
```diff
@@ -0,0 +1,58 @@
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+public class TableProcessorNode<K, V> extends StreamsGraphNode {
```
Created to represent `KTable` operations (`filter`, `transformValues`, `mapValues`)

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:
```diff
@@ -0,0 +1,148 @@
+                                      topicName);
+
+            topologyBuilder.addProcessor(processorParameters.processorName(), processorParameters.processorSupplier(), sourceName);
+
```
For now, this includes the recent changes from @guozhangwang (PR #5163) for optional re-use of source topic as changelog topic.  This optimization will get folded into this PR in a follow-up push.

guozhangwang replied:
I cannot tell how this includes the source topic reuse logic, could you explain a bit?

bbejeck reviewed (commented):
Some comments about the updates here

bbejeck commented:
@guozhangwang @mjsax @vvcephei 

Apologies for the massive PR but it seems unavoidable at this point.  More updates to come to this PR soon, but it's worth reviewing the graph object types and the overall physical plan generation

bbejeck commented:
failure unrelated.

retest this please

bbejeck committed:
Get optimizations part way working

bbejeck committed:
Get repartitioned optimizations working YEAH! refactored repartitioned nodes due differences in building

bbejeck committed:
Get repartitioned optimizations working YEAH! refactored repartitioned nodes due differences in building

bbejeck committed:
only create repartition node if required

bbejeck committed:
KAFKA-6761: break repartitioning and stateful operation into two graph nodes

bbejeck committed:
KAFKA-6761: change name of boolean indicating key changing operation, add check for property to use to optimize, some descriptive comments around optimization

bbejeck committed:
KAFKA-6761: add remove single child method, added system test

bbejeck committed:
KAFKA-6761: KStreamImpl don't need to create dummy graph node for groupByKey just pass along existing parent node

bbejeck committed:
KAFKA-6761: Improved optimization on repartition to handle operators in arbitrary places in topology, not just immediately adjacent to key-changing operation, leave other operations in place if not repartitioning

bbejeck committed:
KAFKA-6761: improved variable names, added better comments, improved find parent functionality, removed unnecessary code

bbejeck committed:
KAFKA-6761: use variable name for serdes

bbejeck committed:
KAFKA-6761: add unit test for optimized and non-optimized topology generation

bbejeck commented:
updated this

bbejeck committed:
KAFKA-6761: rebased from trunk

bbejeck commented:
rebased from trunk

bbejeck committed:
KAFKA-6761: get joins working with repartition optimization, removed some unused code, add reciprocal operations when adding or removing child nodes

guozhangwang commented:
@bbejeck Could you rebase this PR?

bbejeck committed:
KAFKA-6761: fixing rebase errors

bbejeck commented:
retest this please

bbejeck commented:
failure unrelated 

retest this please

bbejeck committed:
KAFKA-6761: change name of graph node field in AbstractStream

bbejeck committed:
KAFKA-6761: refactored StatelessProcessorNode to ProcessorNode, fixed joins by wiring streams source into graph

bbejeck committed:
KAFKA-6761: don't need to pass in graph nodes, just processor information

bbejeck committed:
KAFKA-6761: refactored test to delete and create topics from method, created return object for verification

bbejeck committed:
KAFKA-6761: extracted test into parameterized form

bbejeck committed:
KAFKA-6761: updated integration test to compare results of optimized and non-optimized, verify repartition topic count and include joins

bbejeck commented:
Fixed issue with optimizing joins, updated integration test to include comparing results between optimized and non-optimized results and included a join in the test case.

bbejeck committed:
KAFKA-6761: clean up test

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java:
```diff
@@ -85,18 +71,15 @@ public V apply() {
         if (!sourceName.equals(this.name)) {
             StreamsGraphNode repartitionNode = repartitionNodeBuilder.build();
             streamsGraphNode.addChildNode(repartitionNode);
+            builder.maybeAddNodeForOptimizationMetadata(repartitionNode);
```
It seems like this is happening twice; once inside `addChildNode` and once outside it.

(also applies to other occurrences)

guozhangwang replied:
+1

bbejeck replied:
This is intentional. 
Today, when a repartition is required, the `InternalTopologyBuilder` creates a repartition operation and immediately writes it to the physical plan.  So we need to capture the repartition as a graph node and pass the new node to the `InternalStreamsBuilder` for possible metadata collection for optimization.  

In other cased graph nodes are created in classes that don't subclass the `AbstractStream`.

guozhangwang replied:
Could you explain to us a bit more? I'm still scratching my head now on when do we call 

`AbstractStream#addGraphNode` 

v.s. 

`parentNode.addChildNode; builder.maybeAddNodeForOptimizationMetadata(repartitionNode);`

Because their logic are just the same, right?

bbejeck replied:
Yes, the logic is the same.  But `GroupedStreamAggregateBuilder` does not subclass `AbstractStream` so we need to make those two calls separately.

Secondly, in the example above we are building 2 graph nodes.  The repartition node and the stateful processor node for the aggregation.  

Today when we need to repartition for an aggregation, the `InternalTopologyBuilder` creates a repartition operation and it becomes the parent of the aggregation operation.  So the repartition node needs to be a child of the current `StreamsGraphNode` in the `GroupedStreamAggregateBuilder` but it must be the parent of the aggregation graph node. 

Does this make sense?

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -16,35 +16,59 @@
     protected final StreamsGraphNode root = new StreamsGraphNode(TOPOLOGY_ROOT, false) {
         @Override
-        public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
+        public void writeToTopology(InternalTopologyBuilder topologyBuilder) {
```
nit: could be final ;)

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -16,35 +16,59 @@
+    private final NodeIdComparator nodeIdComparator = new NodeIdComparator();
+    private final Map<StreamsGraphNode, Set<OptimizableRepartitionNode>> keyChangingOperationsToOptimizableRepartitionNodes = new HashMap<>();
+    private final Map<StreamsGraphNode, StreamSinkNode> stateStoreNodeToSinkNodes = new HashMap<>();
+    private final Set<TableSourceNode> tableSourceNodes = new LinkedHashSet<>();
```
I think these are never referenced. Is there still more to do?

guozhangwang replied:
If they are used for the optimization rules that are to be added later, we should remove them from this PR. I'm now not so sure if @bbejeck wants to add the optimizations in this PR, but I'd suggest we do it in a forth one given the current PR is pretty large already.

bbejeck replied:
Removed from this PR and delayed to 4th PR with optimization

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -36,36 +36,35 @@
+
     }
 
     // This method allows to expose the InternalTopologyBuilder instance
```
This is a bit beside the point, but do we actually need this? It's unused in our codebase. Do we expect users to subclass `AbstractStream`? I think the question is applicable; I'm still hoping we could finish disentangling the InternalTopologyBuilder from the StreamsGraph and InternalStreamsBuilder.

vvcephei replied:
Also, on Line 76, we're taking note of which nodes need to be copartitioned. Do we need to capture this information in the StreamsGraph as well?

vvcephei replied:
Aside from these two usages, the only other purpose that the internalTopologyBuilder serves in the StreamsBuilder hierarchy is to add state stores. I think that if we add this to the logical plan first, then we could completely decouple the InternalTopologyBuilder from the InternalStreamsBuilder.

guozhangwang replied:
This is added for users that want to extend `AbstractStream`, one use case of it is https://github.com/fhussonnois/kafkastreams-cep.

Regarding `InternalTopologyBuilder` and `InternalStreamsBuilder`: we need to pass in `InternalTopologyBuilder` into the `StreamsGraphNode` because their `writeToTopology` (this is the one that translates the logical node into one or more physical nodes) needs it, and those graph nodes are accessed from the `InternalStreamsBuilder`, so I think we still need to let it hold a reference of the `InternalTopologyBuilder` anyways.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
     }
 
+    public void buildAndOptimizeTopology(Properties props) {
+        if (!topologyBuilt) {
```
Might be nice to throw an exception if this check fails. It would clearly be a mistake, and it might be nicer for it to break than to do nothing. (I'm specifically thinking if the builder is created, built, modified, and built again; the second modification would just be lost.)

guozhangwang replied:
If users call `StreamsBuilder#build()` multiple times, which we cannot forbid programmatically, we should still only call this function once; on the other hand, if users add a few more operations into the streams topology and then call `build()` again we should probably re-run optimization and generate a new topology, i.e.:

```
1. stream1 = builder.stream(..); stream1.groupby().aggrege();...
2. topology1 = builder.build();   // generate physical plan from the logical plan once.
3. topology2 = builder.build();   // nothing changed, we should return the same topology, hence making sure all the optimizations are deterministic.
4. stream2 = builder.stream(..); stream1.groupby().aggrege();...
5. topology1 = builder.build();   // generate a new physical plan from the logical plan again
```

bbejeck replied:
That was the intent, build once then subsequent calls return the same physical plan from rebuilding.  

As for updates, I'm thinking the use case would be to build the topology incrementally and call build at the end versus incremental calls to build,  in which case the current approach still works.

guozhangwang replied:
Hmm.. for step 4) / 5), now the `topologyBuilt` would still be true and hence we would return the same topology, but that is incorrect right?

bbejeck replied:
What I was thinking of was the developer would build an initial topology with the last statement being the `builder.build()...` Then at that point execute the program and observe the printed topology.  Then go back and update the topology again and run the program a separate time and watch the results.

But thinking about it more, that is an opinionated/subjective view of how to develop, and we should not restrict to one style.  I'll put something to detect if the topology has changed then rebuild if true.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -16,35 +16,59 @@
+import java.util.function.Predicate;
 import java.util.regex.Pattern;
 
 public class InternalStreamsBuilder implements InternalNameProvider {
```
Nit: many variables in here can be final.

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -16,35 +16,59 @@
+    private final AtomicInteger nodeIdCounter = new AtomicInteger(0);
+    private final NodeIdComparator nodeIdComparator = new NodeIdComparator();
+    private final Map<StreamsGraphNode, Set<OptimizableRepartitionNode>> keyChangingOperationsToOptimizableRepartitionNodes = new HashMap<>();
+    private final Map<StreamsGraphNode, StreamSinkNode> stateStoreNodeToSinkNodes = new HashMap<>();
```
This field is never used.

bbejeck replied:
removed until 4th PR with optimization

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+                LOG.debug("Optimizing the Kafka Streams graph");
+                optimize();
+            }
+            final PriorityQueue<StreamsGraphNode> graphNodePriorityQueue = new PriorityQueue<>(5, nodeIdComparator);
```
I wasn't sure why the `NodeIdComparator` implements `Serializable`. If it doesn't need to, then you can get rid of that class and that field and just do:
```
final PriorityQueue<StreamsGraphNode> graphNodePriorityQueue = new PriorityQueue<>(5, Comparator.comparing(StreamsGraphNode::id));
```

guozhangwang replied:
+1, if we are not going to write / serialize the logical plan anywhere we do not need to do that.

bbejeck replied:
This is required by findbugs.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -36,36 +36,35 @@
-        if (parentGraphNode != null) {
-            parentGraphNode.addChildNode(newNode);
-        }
+        streamsGraphNode.addChildNode(newNode);
```
If we remove the `null` check, should we add one in the constructor?

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -16,35 +16,59 @@
     protected final StreamsGraphNode root = new StreamsGraphNode(TOPOLOGY_ROOT, false) {
         @Override
-        public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
+        public void writeToTopology(InternalTopologyBuilder topologyBuilder) {
```
Why remove `final`?

bbejeck replied:
ack merge mistake

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
                        stateUpdateSupplier);
     }
 
+    public void buildAndOptimizeTopology(Properties props) {
```
nit: add `final`

bbejeck replied:
ack. Removed for now as part 3 is for writing physical plan using graph. Part 4 will contain optimization only

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+                final StreamsGraphNode streamGraphNode = graphNodePriorityQueue.remove();
+
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("about write topology for {} child nodes {}", streamGraphNode, streamGraphNode.children());
```
Don't understand the log statement? Something missing there?

bbejeck replied:
same as above will clean up in part 4

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+
+                streamGraphNode.writeToTopology(internalTopologyBuilder);
+
+                for (StreamsGraphNode graphNode : streamGraphNode.children()) {
```
nit: add `final`

bbejeck replied:
same as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+    private void mayOptimizeRepartitionOperations() {
+        StreamsGraphNode optimizedSingleRepartition = null;
+
+        for (Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> streamsGraphNodeSetEntry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
```
nit: add `final` (seems some more in the next lines)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+
+        for (Map.Entry<StreamsGraphNode, Set<OptimizableRepartitionNode>> streamsGraphNodeSetEntry : keyChangingOperationsToOptimizableRepartitionNodes.entrySet()) {
+
+            StreamsGraphNode keyChangingNode = streamsGraphNodeSetEntry.getKey();
```
nit: move after the following `if` (we not' need to get the `keyChangingNode` if we `continue`

bbejeck replied:
ack, will clean up in part 4

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+                                                               parentSourceNode.keySerde(),
+                                                               parentSourceNode.valueSerde());
+
+            // re-use parent id to make sure StreamsGraphNode is evaluated before downstream nodes
```
which "StreamsGraphNode" ?

bbejeck replied:
The repartition node just created, will clarify in part 4

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -131,9 +130,9 @@ public String queryableStoreName() {
         final boolean shouldMaterialize = materializedInternal != null && materializedInternal.isQueryable();
 
         KTableProcessorSupplier<K, V, V> processorSupplier = new KTableFilter<>(this,
```
Super minor nit: I thought the code style says to always put arguments on a new line when splitting args over multiple lines. I only bring this up because I've been doing it that way...

bbejeck replied:
I try to follow that as well, but maybe I'm missing something because I thought the args were all on one line here.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java:
```diff
@@ -17,24 +17,25 @@
 
     private StreamsGraphNode parentNode;
-    private final Collection<StreamsGraphNode> childNodes = new LinkedHashSet<>();
+    private LinkedHashSet<StreamsGraphNode> childNodes = new LinkedHashSet<>();
```
final?

bbejeck replied:
ack

vvcephei reviewed (commented):
Looks good overall to me. I did not go through and prove the correctness of each node, but I can if desired.

I asked a couple of questions.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -403,6 +384,7 @@ public void to(final String topic) {
         to(topic, Produced.<K, V>with(null, null, null));
     }
 
+    @SuppressWarnings("unchecked")
```
Why do we need this here?

bbejeck replied:
ack removed

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:
```diff
@@ -57,7 +59,29 @@
     @Override
     public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
-        //TODO will implement in follow-up pr
+        topologyBuilder.addProcessor(thisProcessorParameters().processorName(),
```
nit: can we use variable instead of getting the names multiple times?

guozhangwang replied:
+1

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorNode.java:
```diff
@@ -28,7 +28,7 @@
  *
  */
-public class StatelessProcessorNode<K, V> extends StreamsGraphNode {
+public class ProcessorNode<K, V> extends StreamsGraphNode {
```
Why this renaming?

bbejeck replied:
During work on this PR, I started to consider this name was a better fit if you insist I can revert.

mjsax replied:
I am 51:49 to keep the old name (so, I don't insist in reverting). \cc @guozhangwang @vvcephei WDYT?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorNode.java:
```diff
@@ -40,36 +40,42 @@
         this(nodeName, processorParameters, repartitionRequired);
 
-        this.multipleParentNames = multipleParentNames;
+        this.multipleParentNames = new ArrayList<>(multipleParentNames);
```
nit: simplify `multipleParentNames` -> `parentNames` (`Names` is already plural, should be good enough).

Im wondering, it it might be better to track the parents in `StreamsGraphNode` ?

bbejeck replied:
ack on the name.

For now, I'd prefer to leave `parentNames` in the `ProcessorNode` as we specifically add multiple parent names in `KStream#merge`. Or maybe refactor always to take a list `parentNames` that would simplify the logic 

EDIT: I take that back, we want to rely on getting the parent name for the current graph node by calling `parentNode.name()` the multiple parent names comes from the case mentioned above and are needed for the `InternalStreamsBuilder` to complete the merge processor,  so I'll leave as is for now.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorNode.java:
```diff
@@ -40,36 +40,42 @@
     @Override
     public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
-        //TODO will implement in follow-up pr
+        if (processorParameters != null) {
```
When could this be `null`?

bbejeck replied:
I've removed this line, left over from a previous refactoring.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorNode.java:
```diff
@@ -40,36 +40,42 @@
     public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
-        //TODO will implement in follow-up pr
+        if (processorParameters != null) {
+            if (multipleParentNames.isEmpty()) {
```
When would this be not empty?

bbejeck replied:
Only used for merge node, so it's empty most of the time. I'll look to see if I can refactor internally.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StatefulProcessorNode.java:
```diff
@@ -42,18 +39,24 @@ public StatefulProcessorNode(final String nodeName,
+        final String processorName = processorParameters().processorName();
+        final ProcessorSupplier processorSupplier = processorParameters().processorSupplier();
+
+        topologyBuilder.addProcessor(processorName, processorSupplier, parentNode().nodeName());
```
Why doe we assume, that a stateful operator has always one parent? What about a materialized table-table join?

bbejeck replied:
table-table joins are represented in a separate node, `KTableKTableJoinNode`.  I used a specific node for this case IMHO there is too much information needed in the table-table join to generalize.

guozhangwang replied:
Just to be general enough, we should just loop over all parent nodes from the list and add all their node names here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSinkNode.java:
```diff
@@ -41,37 +41,19 @@ public StreamSinkNode(final String nodeName,
+        final Serializer<V> valSerializer = producedInternal.valueSerde() == null ? null : producedInternal.valueSerde().serializer();
+        final StreamPartitioner<? super K, ? super V> partitioner = producedInternal.streamPartitioner();
+
+
```
nit: remove empty line

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamStreamJoinNode.java:
```diff
@@ -45,48 +44,34 @@
     public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
-        //TODO will implement in follow-up pr
+
+        topologyBuilder.addProcessor(thisProcessorParameters().processorName(), thisProcessorParameters().processorSupplier(), thisWindowedStreamProcessorParameters.processorName());
```
nit: introduce variable to get name only once?

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java:
```diff
@@ -29,37 +28,33 @@
-        isGlobalKTableJoin = globalKTableJoin;
-    }
+        // Stream - GlobalKTable join
+        topologyBuilder.addProcessor(processorName, processorSupplier, parentNode().nodeName());
```
This is done for both join, right? (ie, update comment?)

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java:
```diff
@@ -29,37 +28,33 @@
-    boolean isGlobalKTableJoin() {
-        return isGlobalKTableJoin;
-    }
+        // Steam - Table join
```
nit: `Steam - Table join [only]`

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java:
```diff
@@ -47,10 +48,26 @@ public void setParentNode(final StreamsGraphNode parentNode) {
 
+    public void clearChildren() {
+        for (StreamsGraphNode childNode : childNodes) {
+            if (childNode.parentNode != null && childNode.parentNode.equals(this)) {
```
Why do we need this check? It seem that `parentNode` cannot be `null`, and that it would be `this` always, too? (assuming one parent node, what is not generic)

bbejeck replied:
Since the optimization is pushed out to a 4th PR, this method is removed.  I'll clean up in 4th PR. 

The original idea was to prevent any errors by calling `clearChildren` _*after*_ the children of a given parent node have migrated to another parent.  Probably better to eliminate the checks and document the calling order when updating the graph.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java:
```diff
@@ -47,10 +48,26 @@ public void setParentNode(final StreamsGraphNode parentNode) {
+    }
+
+    public boolean removeChild(StreamsGraphNode child) {
+        if (child.parentNode != null && child.parentNode.equals(this)) {
```
as above.

bbejeck replied:
same comment from above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java:
```diff
@@ -17,19 +17,21 @@
+import org.apache.kafka.streams.state.KeyValueStore;
 
-public class TableProcessorNode<K, V, S extends StateStore> extends StreamsGraphNode {
+public class TableProcessorNode<K, V> extends StreamsGraphNode {
```
Why not extend `StatefulProcessorNode` ? Why remove the generic type? Can't it be windowed, too? What about custom stores in `transform()` ?

guozhangwang replied:
+1.

bbejeck replied:
I removed the generic type from the class definition and changed the `materializedInternal` from `S` to `KeyValueStore<Bytes, byte[]` as it matches the `MaterializedInternal` parameter used in `KTable#filter`, `KTable#mapValues`, `KTable#transformValues`.  I can revert that if you want.

Looking at `StatefulProcessorNode` it's only used from `KStream` for `process` and `transform`, but those methods never pass a strore builder or materialized, just store names.

 What makes sense to me is to refactor `StatefulProcessorNode` to remove the store builder then have `TableProcessorNode` extend `StatefulProcessorNode` . WDYT?

I'm not sure what you mean custom stores in `transform()`, `KStream#transform` takes a list of store names which is captured in the `StatefulProcessorNode` is that what you are referring to?

bbejeck replied:
EDIT: I've put the generic type back, but requires a cast internally, you have to let me know what you think

mjsax replied:
Ack. Makes sense that `transform` does not apply here as this is a `*Table*ProcessorNode`.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java:
```diff
@@ -42,6 +44,15 @@ public TableProcessorNode(final String nodeName,
 
     @Override
     public void writeToTopology(InternalTopologyBuilder topologyBuilder) {
```
nit: add `final`

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:
```diff
@@ -30,7 +29,7 @@
 public class TableSourceNode<K, V> extends StreamSourceNode<K, V> {
 
-    private StoreBuilder<KeyValueStore<K, V>> storeBuilder;
+    private StoreBuilder storeBuilder;
```
Would it be better to put a generic `<S>` her and change to `TableSourceNode<K, V, S extends StateStore>`? Also, should we take `builder.windowedTable()` into account already? Pretty sure the KIP will be accepted.

bbejeck replied:
Ack, required some other minor changes, you'll have to let me know what you think

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:
```diff
@@ -75,7 +62,31 @@ boolean isGlobalKTable() {
     @Override
     public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
-        //TODO will implement in follow-up pr
+        String topicName = getTopicNames().get(0);
```
nit: `final`

bbejeck replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:
```diff
@@ -84,7 +95,7 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
         private String sourceName;
         private String topic;
         private ConsumedInternal<K, V> consumedInternal;
-        private StoreBuilder<KeyValueStore<K, V>> storeBuilder;
```
as above

guozhangwang replied:
Same here.

mjsax reviewed (commented):
One initial pass -- did not look at the tests yet.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+                  optimizedSingleRepartition != null ? optimizedSingleRepartition.children() : null);
+    }
+
+    private StreamsGraphNode createRepartitionNode(String name, Serde keySerde, Serde valueSerde) {
```
KGroupedTableImpl also have a duplicated `createRepartitionNode`.

bbejeck replied:
Ack will try to collapse the two in 4th PR, but for now, the `createReparitionNode` is removed from `InternalStreamsBuilder` until 4th PR

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+        mayOptimizeRepartitionOperations();
+    }
+
+    private void mayOptimizeRepartitionOperations() {
```
May need some more explanation of this optimization rule.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+    }
+
+    private void optimize() {
+        mayOptimizeRepartitionOperations();
```
Meta comment: if we are going to add more optimization rules in `optimize()`, should we keep this PR as a plain one that do not enforce any optimizations, so that we can then consider each function separately, that helps more concentrated reviews and reduce large PR burdens.

bbejeck replied:
Ack removed optimization in favor of pushing a 4th PR with optimizations only once 3rd PR is merged.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamStreamJoinNode.java:
```diff
@@ -45,48 +44,34 @@
                          final StoreBuilder<WindowStore<K, V1>> thisWindowStoreBuilder,
                          final StoreBuilder<WindowStore<K, V2>> otherWindowStoreBuilder,
-                         final Joined<K, V1, V2> joined,
-                         final String leftHandSideStreamName,
```
This is nice cleanup.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:
```diff
@@ -18,12 +18,14 @@
 package org.apache.kafka.streams.kstream.internals.graph;
 
 import org.apache.kafka.streams.kstream.ValueJoiner;
+import org.apache.kafka.streams.kstream.internals.KeyValueStoreMaterializer;
```
Meta comment: seems we have migrated the node classes in a previous PR so could you update the description of this PR to remove the statement that we changed the package here?

bbejeck replied:
ack updated

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java:
```diff
@@ -94,4 +107,12 @@ public InternalStreamsBuilder internalStreamsBuilder() {
     public abstract void writeToTopology(final InternalTopologyBuilder topologyBuilder);
 
+    @Override
+    public String toString() {
```
Should we override this for all the subclasses so that we can have a more informative intermediate logical plan representation for debugging purposes?

bbejeck replied:
yep, I'll have to put some thought into what's relevant without being too spammy

bbejeck committed:
KAFKA-6761: added to test, removed unnecessary log statement, better method name

bbejeck commented:
>I cannot tell how this includes the source topic reuse logic, could you explain a bit?

In this PR we're still relying on the fix put into 2.0, in 4th PR we'll revert that change and include it along with the repartition optimization.

bbejeck commented:
>+1, if we are not going to write / serialize the logical plan anywhere we do not need to do that.

required to keep find-bugs happy

bbejeck committed:
updates per comments

bbejeck commented:
updated this

guozhangwang commented:
@bbejeck I made another pass over the updated PR, and left two follow-up comments as in https://github.com/apache/kafka/pull/5201#discussion_r199012496 and https://github.com/apache/kafka/pull/5201/files#r199014145.

bbejeck commented:
retest this please

bbejeck commented:
retest this please

guozhangwang commented:
Regarding the meta comment about `AbstractStream#addGraphNode`, now I understand the duplicate logic is because `KStreamAggregateProcessorBuilder` is not `AbstractStream` and hence cannot use this function.

My personal preference would then be, moving the `addGraphNode` from `AbstractStream` to `InternalStreamsBuilder`, with a small change on API to require passing in both the child and the parent node reference. In this case 1) beyond `AbstractStream` we can still call it, and 2) moving forward if we need to have more then one StreamGraphNode within an AbstractStream it would be more natural to do so.

guozhangwang commented:
@bbejeck please let us when this PR is ready for review again.

bbejeck committed:
KAFKA-6761: update tests from trunk rebase

bbejeck committed:
KAFKA-6761: move addGraphNode to InternalStreamsBuilder from AbstractStream update KStreamImpl

bbejeck committed:
KAFKA-6761: finished moving addGraphNode to InternalStreamsBuilder from AbstractStream updated GroupedSteamAggregateBuilder, KGroupedTableImpl, KStreamImpl, KTableImpl, and AbstractStreamTest

bbejeck committed:
KAFKA-6761: remove condition for only calling build topology once added test

bbejeck commented:
updated this.  Moved `addGraphNode` from `AbstractStream` to `InternalStreamsBuilder`.  Also removed the restriction of only building topology only once and added a unit test for it.

Also rebased with trunk.

Ready for reviews again

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +203,53 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+                LOG.debug("Adding nodes to topology {} child nodes {}", streamGraphNode, streamGraphNode.children());
+            }
+
+            if (!streamGraphNode.hasWrittenToTopology()) {
```
For removing the restriction of only building topology once, I've decided on the "traditional" approach when traversing a graph to only visit (in this case write its contents to the `InternalTopologyBuilder`) graph nodes not already visited.

I feel a similar approach should work with optimization. But since this PR does not include the optimization, I'd prefer to defer any discussions on what will and won't work for multiple `build()` calls until we have the PR for applying the optimization pushed.

guozhangwang replied:
Ack, lgtm.

bbejeck commented:
just remembered, need to update `toString` for individual graph nodes

bbejeck committed:
KAFKA-6761: marking variables as final

bbejeck committed:
KAFKA-6761: adding initial toString to all graph nodes

bbejeck commented:
updated this 

address last comments, rebased from trunk

bbejeck committed:
KAFKA-6761: cherry-pick https://github.com/apache/kafka/pull/5362

bbejeck committed:
KAFKA-6761: cherry-picked https://github.com/apache/kafka/pull/5360

bbejeck committed:
KAFKA-6761: cherry-picked https://github.com/apache/kafka/pull/5363

bbejeck committed:
KAFKA-6761: cherry-pick https://github.com/apache/kafka/pull/5364

bbejeck commented:
retest this please

bbejeck commented:
updated to include clean-up from @vvcephei on MINOR cleanup PRs

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:
```diff
@@ -103,7 +96,8 @@ public VR apply(final K readOnlyKey, final V value) {
     }
 
-    static <K, V, VR> ValueTransformerWithKeySupplier<K, V, VR> toValueTransformerWithKeySupplier(final ValueTransformerSupplier<V, VR> valueTransformerSupplier) {
+    static <K, V, VR> ValueTransformerWithKeySupplier<K, V, VR> toValueTransformerWithKeySupplier(
```
nit: you can configure the IDE to turn auto-newline-beyond-column-limit off :)

bbejeck replied:
Ack

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/GroupedStreamAggregateBuilder.java:
```diff
@@ -76,27 +63,25 @@ public V apply() {
         if (!sourceName.equals(this.name)) {
             StreamsGraphNode repartitionNode = repartitionNodeBuilder.build();
-            streamsGraphNode.addChildNode(repartitionNode);
+            builder.addGraphNode(this.streamsGraphNode, repartitionNode);
```
nit: this.streamsGraphNode can be replaced with parentNode.

bbejeck replied:
ack

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java:
```diff
@@ -47,7 +49,7 @@ public void setParentNode(final StreamsGraphNode parentNode) {
     }
 
-    public Collection<StreamsGraphNode> children() {
+    public Set<StreamsGraphNode> children() {
```
Why change Collection to Set? The former is more generalized right?

bbejeck replied:
Can't recall, I think it was related to changes for optimization implementation, I'll revert for now.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -228,8 +203,53 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+
+    void maybeAddNodeForOptimizationMetadata(final StreamsGraphNode node) {
+        node.setId(nodeIdCounter.getAndIncrement());
+        node.setInternalStreamsBuilder(this);
```
The `StreamsGraphNode#internalStreamsBuilder()` seems not used anywhere?

bbejeck replied:
ack, removed

EDIT: put back for now to avoid findbugs error.

guozhangwang replied:
Actually, my point is that the private `internalStreamsBuilder` field is not used anywhere either, so we can remove both this field as well as the getter function.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -423,49 +432,37 @@ public void to(final TopicNameExtractor<K, V> topicExtractor, final Produced<K,
 
     @SuppressWarnings("unchecked")
-    private void to(final TopicNameExtractor<K, V> topicExtractor, final ProducedInternal<K, V> produced) {
+    private <KInner> void to(final TopicNameExtractor<K, V> topicExtractor, final ProducedInternal<K, V> produced) {
```
I know it is cherry-picked from @vvcephei 's PR, but it seems we drops this type information in the logical StreamsGraphNode anyways. Maybe @vvcephei can comment whether we still need this?

vvcephei replied:
Yeah, in this case, the type parameter is unused, and we're also suppressing warnings. I think we can either ditch the parameter or remove the suppression.

FWIW, I have a todo after this PR is merged to make another pass to eliminate warnings and unnecessary suppressions, so feel free to just drop the type parameter, and I'll figure out whether the suppression is necessary later on. So as to not expand this PR further.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -423,49 +432,37 @@ public void to(final TopicNameExtractor<K, V> topicExtractor, final Produced<K,
-        }
 
-        return new KStreamImpl<>(builder, name, sourceNodes, true, null);
+        return new KStreamImpl<>(builder, name, sourceNodes, true, transformNode);
```
Why we pass `null` before and now we need to pass in the `transformNode`? Was it a bug before and we fixed it here?

vvcephei replied:
We were never actually using the graph before, right? So a missed node might go unnoticed until now, when we actually build the topology from the graph.

bbejeck replied:
looking from the history looks like it was a bug but fixed here.

guozhangwang replied:
Thanks for confirming :) just want to make sure it is not a regression.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -488,77 +485,69 @@ private void to(final TopicNameExtractor<K, V> topicExtractor, final ProducedInt
 
-        StatefulProcessorNode<K, V> transformNode = new StatefulProcessorNode<>(name,
+        final ProcessorParameters processorParameters = new ProcessorParameters<>(processorSupplier, name);
+        final StatefulProcessorNode<K, V> transformNode = new StatefulProcessorNode<>(name,
```
nit: `processNode`.

bbejeck replied:
ack

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java:
```diff
@@ -29,37 +30,42 @@
-    boolean isGlobalKTableJoin() {
-        return isGlobalKTableJoin;
-    }
+        // Steam - Table join only
```
The comments here are a bit confusing: line 63 is for both globaltable-stream join and table-stream join, and line 66 is for table-stream join only.

bbejeck replied:
ack, updated

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -800,43 +788,37 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
         Objects.requireNonNull(serialized, "serialized can't be null");
         final SerializedInternal<KR, V> serializedInternal = new SerializedInternal<>(serialized);
-        final StatelessProcessorNode<K, V> graphNode = internalSelectKey(selector);
+        final ProcessorNode<K, V> graphNode = internalSelectKey(selector);
```
nit: selectKeyMapNode.

bbejeck replied:
ack

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -800,43 +788,37 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
                                         serializedInternal.valueSerde(),
                                         this.repartitionRequired,
-                                        graphNode);
+                                        streamsGraphNode);
```
We did not add a new logical node here anymore in this PR, is this intentional?

bbejeck replied:
Yes, it's intentional. What I found during testing is that we don't need to create a node here.

I initially had a logical node at this point, but it never rendered any details for the physical plan, as it's methods on the `KGroupedStreamImpl` that provide details for the next operation of the physical plan, thus we only need to create a new logical node when one of those operations are specified.

When I did create a new logical plan node here, it contained no details to render, so I needed to put in checks for `null` `ProcessorParameters`. Having a placeholder node was somewhat probalmatic, so instead of a "dummy" node which I found to be confusing as well, I removed creating a new node at this point, and IMHO is better off this way.

guozhangwang replied:
Thanks!

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -861,11 +843,14 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
         KStreamImplJoin(final boolean leftOuter,
-                        final boolean rightOuter) {
+                        final boolean rightOuter,
+                        final StreamsGraphNode graphNode) {
```
nit: rename to `parentGraphNode`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -874,71 +859,83 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
+            );
+
+            final ProcessorParameters thisWindowStreamProcessorParams = new ProcessorParameters(thisWindowedStream, thisWindowStreamName);
+            final ProcessorNode<K1, V1> thisWindowedStreamsNode = new ProcessorNode<>(thisWindowStreamName, thisWindowStreamProcessorParams);
```
These four graph nodes: thisWindowedStreamsNode, thisStreamsGraphNode, otherWindowedStreamsNode, otherStreamsGraphNode, seems not needed any more since we now use an "umbrella" joinGraphNode that contains all the information, right? Ditto for other join implementations.

If it is indeed the case, then I'm wondering why our unit test does not fail, as it will unnecessarily add more processor nodes, and hence should cause some unit test to fail.

bbejeck replied:
@guozhangwang  the graph node is indeed required, and the tests are passing correctly.  Implementing the join for optimization was possibly the most "intricate" detail to get correct.

Previously when we generated the entire physical plan upfront, the parent names for the two `KStreamJoinWindow` instances were generated inside the `KStreamImplJoin.join` method and passed to the InternalStreamsBuilder to wire up the processors.  

The first pass of the `StreamStreamJoinNode` followed this approach, and everything was wired together in one operation.  But the problem is with that approach we lose the ability to optimize any join nodes much for the same reason as before, we write everything required for the join to the `InternalTopologyBuilder` and lose any concept if optimization has occurred and what the new parent node is. 

 Either stream involved in a join can be eligible for optimization concerning repartitioning.  In that case, we need to be able to use the updated parent node names for either or both of the `KStreamJoinWindow` instances involved in the join.

So what has been done here is to add the two `KStreamJoinWindow` instances explicitly as child nodes of the two `KStream` instances passed as parameters from the `doJoin` method.  The key point here is that one or both of the original `KStream` instances may have required a repartitioning.  

This fact requires us to explicitly attach the `KStreamJoinWindow` as a child node of the passed in `KStream` instance so if a repartition optimization does occur; the correct parent name is used by the `KStreamJoinWindow` when the physical plan is written.

As for your question of adding too many processors, when traversing the graph the nodes representing the `KStreamJoinWindow` processors have their details written, but in the `StreamStreamJoinNode` the previous calls writing the details for the `KStreamJoinWindow` processors out for the physical plan have been removed. Thus the number of processors written out in the physical plan is correct.

guozhangwang replied:
My question is that, in this function we call the following `addGraphNode` calls:

```
builder.addGraphNode(thisStreamsGraphNode, thisWindowedStreamsNode);
builder.addGraphNode(otherStreamsGraphNode, otherWindowedStreamsNode);
builder.addGraphNode(this.parentGraphNode, joinGraphNode);
```

And in the returned `KStreamImpl` we passed in `joinGraphNode`. Note the `parentGraphNode` is passed as the latest node added to the topology for now, and hence it looks like although `thisStreamsGraphNode` and `otherStreamsGraphNode` each added a child, they do not have any parents, and hence these two "branches" are sort-of dangling as they are not connected to the topology graph at all.. Could you maybe explain a bit more on this logic?

bbejeck replied:
>it looks like although thisStreamsGraphNode and otherStreamsGraphNode each added a child, they do not have any parents, and hence these two "branches" are sort-of dangling as they are not connected to the topology graph at all

@guozhangwang  great question.  

In building the topology graph, it's not possible for a graph node to not have a parent.  

It is easier for me to explain the logic best with some examples. 

First, let's consider no repartitioning
```
   KStream<String, String> streamA = builder.stream("topicA");
   KStream<String, String> streamB = builder.stream("topicB");
   KStream<String, String> joined = streamA.join(streamB, vj, JoinWindows.of(5000));
```

In this case we a repartition did not occur in `doJoin`, so the `thisStreamsGraphNode` and `otherStreamsGraphNode`, from the `lhs` and `other` `KStream` instances respectively, represent the graph nodes for `streamA` and `streamB` and the parent node is `root`. Anything created directly from the `StreamBuilder` always has the parent node of `root`.   If a `KStream`  instance is created by another upstream operation like `builder.stream("topicB").filter(...)` , then its parent would be the graph node representing the `filter` operation.

In this case in the `thisStreamsGraphNode` and `parentGraphNode` are the same instance, and it ends up with two child nodes, the `thisWindowedStreamsNode` and the `joinGraphNode`.  I prefer to keep these two nodes separate as we don't need to keep track of repartitioning, I find the logic more clear to use the graph nodes from the `lhs` and `other` `KStream` instances.

Next repartitioning is required:
```
  KStream<String, String> streamA = builder.stream("topicA");
  KStream<String, String> streamB = builder.stream("topicB");
  KStream<String, String> newKeyStream = streamA.selectKey((k,v) -> k+v);
  KStream<String, String> joined = newKeyStream.join(streamB, vj, JoinWindows.of(5000));
```

Now a repartition occurred in `doJoin` and the `thisStreamsGraphNode` is now a repartition graph node, with its parent being the `selectKey` graph node and the grandparent node is the original `streamA` graph node.  

If `streamB` had required a repartition the parent-child structure would be the same, the `otherStreamsGraphNode` would be a repartition node with the parent being the graph node representing the key changing operation and the grandparent node representing the original `streamB` graph node.

So at all times the `thisStreamsGraphNode` and `otherStreamsGraphNode` have parents of either `root` or some other upstream `KStream` operation and the `thisStreamsGraphNode` and `otherStreamsGraphNode` need to add the respective window stream processor graph nodes as children, so the physical plan is rendered correctly.

Does this make sense?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -487,72 +487,72 @@ public V apply(final K key, final Change<V> change) {
-                                                                                      processorParameters,
-                                                                                      false,
-                                                                                      Collections.<String>emptyList());
+        final ProcessorNode<K1, V1> graphNode = new ProcessorNode<>(
```
nit: groupByMapNode.

bbejeck replied:
ack

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/OptimizableRepartitionNode.java:
```diff
@@ -53,33 +56,41 @@
 
+    @Override
+    public String toString() {
+        return "OptimizableRepartitionNode{} " + super.toString();
```
Should be `OptimizableRepartitionNode{ + super.toString() + }`?

vvcephei replied:
If so, then this comment would apply to all the other nodes as well.

bbejeck replied:
ack

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java:
```diff
@@ -17,24 +17,26 @@
 
     private StreamsGraphNode parentNode;
-    private final Collection<StreamsGraphNode> childNodes = new LinkedHashSet<>();
+    private final LinkedHashSet<StreamsGraphNode> childNodes = new LinkedHashSet<>();
```
Ditto here.

bbejeck replied:
ack, reverted

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorNode.java:
```diff
@@ -26,9 +26,8 @@
- *
  */
-public class StatelessProcessorNode<K, V> extends StreamsGraphNode {
+public class ProcessorNode<K, V> extends StreamsGraphNode {
```
nit: Since we have another class named `ProcessorNode` already could we still name it `StatelessProcessorNode`?

vvcephei replied:
I thought it was a little weird previously that `StatefulProcessorNode` extends `StatelessProcessorNode`, since subclassing is generally an "is a" relationship. So it was previously saying "StatefuProcessorNode is a StatelessProcessorNode", which is silly of course.

Maybe we can call it `ProcessorGraphNode` to avoid a collision?

bbejeck replied:
ack

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorNode.java:
```diff
@@ -37,39 +36,61 @@
     // to get registered with InternalStreamsBuilder
 
-    private List<String> multipleParentNames = new ArrayList<>();
+    private List<String> parentNames = new ArrayList<>();
```
The `StreamsGraphNode#parentName` and here `parentNames` relationship is a bit awkward, as the only difference is for `merge`. Could we just use `parentNames` on the `List<String> StreamsGraphNode` directly, and replace `setParentNode` with `addParentNode`. And in all other nodes we just call `addParentNode` only once, while for `merge` we call it twice for each fo the merging streams.

bbejeck replied:
Hmm, I'm not sure. While I agree with you that the `StreamsGraphNode#parentName` and here `parentNames` relationship is a bit awkward but I think changing `setParentNode` to `addParentNode` can be equally as awkward. 

 While I'm up for changing this in some way I'd prefer to leave how we establish the parent-child relationship the same, as we'd be making a cross-cutting change for just the `merge` case, and here it's isolated for just `merge`.


EDIT: I get what you are saying, I'll try to change and see if I can get it to work.

EDIT part II:  Updated and implemented as you suggested, good call.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/ProcessorNode.java:
```diff
@@ -37,39 +36,61 @@
+
+        // merge operation passes in multiple parent names
+        // so if it's empty add the parent node name
+        if (parentNames.isEmpty()) {
```
See my other comment above: we can replace this logic with the `addParentNode`.

bbejeck replied:
replied above

bbejeck replied:
ack, done

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StatefulProcessorNode.java:
```diff
@@ -17,43 +17,58 @@
-        super(nodeName,
-              processorParameters,
-              repartitionRequired);
+        super(
```
I think this coding restyle is really not necessary.

vvcephei replied:
:+1: I think this was from my misinterpretation of the code style.

bbejeck replied:
ack. But I think we need to come together on this as a team.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamSourceNode.java:
```diff
@@ -51,21 +54,54 @@ public StreamSourceNode(final String nodeName,
 
-    public Collection<String> getTopicNames() {
-        return topicNames;
+    public List<String> getTopicNames() {
```
Why changing collection to list?

bbejeck replied:
out of convenience, reverted

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java:
```diff
@@ -17,24 +17,26 @@
     private boolean keyChangingOperation;
     private Integer id;
     private InternalStreamsBuilder internalStreamsBuilder;
+    private boolean hasWrittenToTopology = false;
```
For line 35 above: see my other comment: `internalStreamsBuilder` seems not used anywhere.

bbejeck replied:
I think it's needed for optimization PR, but I'll remove here and will add back if necessary

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java:
```diff
@@ -39,9 +44,28 @@ public TableProcessorNode(final String nodeName,
+        final boolean shouldMaterialize = materializedInternal != null && materializedInternal.isQueryable();
+        final String processorName = processorParameters.processorName();
+
+        topologyBuilder.addProcessor(processorName, processorParameters.processorSupplier(), parentNode().nodeName());
```
Ditto above: if we take the underlying `StreamsGraphNode` to have `list of parentNodes`, then for all such callers we will generalize to loop over all parent nodes and add their node names. Although for now we will only have two parents for `merge`, and one parent for any other types.

bbejeck replied:
ack, updated

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:
```diff
@@ -39,7 +40,7 @@
                     final String topic,
                     final ConsumedInternal<K, V> consumedInternal,
-                    final StoreBuilder<KeyValueStore<K, V>> storeBuilder,
+                    final StoreBuilder storeBuilder,
```
The `StoreBuilder` should be templated as `StoreBuilder<S>`. Ditto elsewhere.

bbejeck replied:
ack

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:
```diff
@@ -59,6 +59,7 @@
     public void shouldAllowJoinUnmaterializedFilteredKTable() {
         final KTable<Bytes, String> filteredKTable = builder.<Bytes, String>table("table-topic").filter(MockPredicate.<Bytes, String>allGoodPredicate());
         builder.<Bytes, String>stream("stream-topic").join(filteredKTable, MockValueJoiner.TOSTRING_JOINER);
+        builder.build();
```
Should we just do `final ProcessorTopology topology = builder.build();` and remove the line below? Same elsewhere.

bbejeck replied:
Actually, we still need both The `StreamsBuilder.build()` call returns a `Topology` while the `InternalTopologyBuilder.build()` returns a `ProcessoryTopology`.   


While the method names seem to be a bit overloaded, the terminology and methods pre-date this PR, so maybe we can do a follow-up PR to look at renaming and clarifying things some.

guozhangwang replied:
Ack, thanks for clarifying.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -228,12 +228,14 @@ public void shouldRestoreToWindowedStores() throws IOException {
 
-        new InternalStreamsBuilder(internalTopologyBuilder)
-            .stream(Collections.singleton("topic"), new ConsumedInternal<>())
+        InternalStreamsBuilder builder = new InternalStreamsBuilder(internalTopologyBuilder);
```
nit: add final.

bbejeck replied:
ack

guozhangwang reviewed (commented):
Made a pass over the PR.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -57,18 +67,10 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
-        StreamSourceNode<K, V> streamSourceNode = new StreamSourceNode<>(name,
+        final StreamSourceNode<K, V> streamSourceNode = new StreamSourceNode<>(name,
                                                                          topics,
                                                                          consumed);
```
nit: alignment is off.

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -77,17 +79,11 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
-        StreamSourceNode<K, V> streamPatternSourceNode = new StreamSourceNode<>(name,
+        final StreamSourceNode<K, V> streamPatternSourceNode = new StreamSourceNode<>(name,
                                                                                 topicPattern,
                                                                                 consumed);
```
nit: alignment is off.

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -158,24 +142,15 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
                                                                                                     .withTopic(topic)
                                                                                                     .withProcessorParameters(processorParameters)
                                                                                                     .isGlobalKTable(true)
                                                                                                     .build();
```
nit: alignment is off.

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:
```diff
@@ -144,7 +128,7 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
-    public <K, V> GlobalKTable<K, V> globalTable(final String topic,
+    public <K, V, S extends StateStore> GlobalKTable<K, V> globalTable(final String topic,
                                                  final ConsumedInternal<K, V> consumed,
                                                  final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materialized) {
```
nit: alignment is off.

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -116,43 +113,43 @@
+        final ProcessorParameters processorParameters = new ProcessorParameters<>(new KStreamFilter<>(predicate, false), name);
+        final ProcessorNode<K, V> filterProcessorNode = new ProcessorNode<>(name,
+                                                                      processorParameters,
+                                                                      repartitionRequired);
```
nit: alignment is off.

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -116,43 +113,43 @@
-                                                                                           repartitionRequired);
+        final ProcessorNode<K, V> filterNotProcessorNode = new ProcessorNode<>(name,
+                                                                         processorParameters,
+                                                                         repartitionRequired);
```
nit: alignment is off.

bbejeck replied:
ack

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -584,61 +573,62 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
 
         return new KStreamImpl<>(builder, repartitionedSourceName, Collections
-            .singleton(repartitionedSourceName), false, repartitionNode);
+            .singleton(repartitionedSourceName), false, optimizableRepartitionNode);
```
This is (and was) a weird line break.

bbejeck replied:
ack fixed, need to update IntelliJ formatting

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:
```diff
@@ -584,61 +573,62 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
-        String sinkName = builder.newProcessorName(SINK_NAME);
-        String nullKeyFilterProcessorName = builder.newProcessorName(FILTER_NAME);
-        String sourceName = builder.newProcessorName(SOURCE_NAME);
+        final Predicate<K1, V1> nullKeyPredicate = (k, v) -> k != null;
```
`notNullKeyPredicate`?

bbejeck replied:
ack

bbejeck commented:
Kicked off system tests https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1879/

Also kicked off streams smoke tests https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1880/

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilderTest.java:
```diff
@@ -255,10 +260,12 @@ public void shouldMapStateStoresToCorrectSourceTopics() {
         final KStream<String, String> mapped = playEvents.map(MockMapper.<String, String>selectValueKeyValueMapper());
         mapped.leftJoin(table, MockValueJoiner.TOSTRING_JOINER).groupByKey().count(Materialized.<String, Long, KeyValueStore<Bytes, byte[]>>as("count"));
+        builder.buildAndOptimizeTopology();
+        assertEquals(Collections.singletonList("table-topic"), builder.internalTopologyBuilder.stateStoreNameToSourceTopics().get("table-store"));
```
duplicate test line?

bbejeck replied:
ack

vvcephei reviewed (commented):
Mostly nits. LGTM overall, modulo some of Guozhang's comments.

bbejeck committed:
KAFKA-6761: committing incrementally, updates per comments part I

bbejeck committed:
KAFKA-6761: per comments re-naming ProcessorNode to StatelessProcessorNode

bbejeck committed:
KAFKA-6761: updates per comments, added ability for multiple parents, cleaned up merge operation and StatelessProcessorNode

bbejeck commented:
>I thought it was a little weird previously that StatefulProcessorNode extends StatelessProcessorNode, since subclassing is generally an "is a" relationship. So it was previously saying "StatefuProcessorNode is a StatelessProcessorNode", which is silly of course.
>Maybe we can call it ProcessorGraphNode to avoid a collision?

I like `ProcessorGraphNode` @guozhangwang @mjsax WDYT?

bbejeck committed:
KAFKA-6761: final round of updating for comments

bbejeck commented:
@guozhangwang @vvcephei updated this and rebased from trunk

guozhangwang commented:
> I like ProcessorGraphNode @guozhangwang @mjsax WDYT?

Sounds good to me.

bbejeck commented:
failure unrelated.

retest this please

bbejeck committed:
KAFKA-6761: rename StatelessProcessorNode, minor cleanup

bbejeck commented:
updated to rename `StatelessProcessorNode` to `ProcessorGraphNode` from comments.

guozhangwang commented:
One more meta comment: we had a bunch of warnings due to lack of types for ProcessorParameters. This can be resolved by adding the following, e.g. in `KStreamImpl#filter()`:

```
final ProcessorParameters<? super K, ? super V> processorParameters = new ProcessorParameters<>(new KStreamFilter<>(predicate, true), name);


final ProcessorGraphNode<? super K, ? super V> filterNotProcessorNode = new ProcessorGraphNode<>(name,
                                                                                         processorParameters,
                                                                                         repartitionRequired);
```

I.e. we can still maintain the typing information here when constructing the logical node.

bbejeck committed:
KAFKA-6761: add ability for multiple parents, some compiler warnings

bbejeck commented:
updated this

bbejeck committed:
KAFKA-6761: missed final in a couple of spots

guozhangwang commented:
LGTM. Merging to trunk.
